source/extensions/filters/http/ext_proc/ext_proc.h

```diff
@@ -278,10 +278,16 @@ class Filter : public Logger::Loggable<Logger::Id::ext_proc>,
   void onMessageTimeout();
   void onNewTimeout(const ProtobufWkt::Duration& override_message_timeout);
 
-  void sendBodyChunk(ProcessorState& state, const Buffer::Instance& data,
-                     ProcessorState::CallbackState new_state, bool end_stream);
+  envoy::service::ext_proc::v3::ProcessingRequest
+  setupBodyChunk(ProcessorState& state, const Buffer::Instance& data, const bool end_stream);
+  void sendBodyChunk(ProcessorState& state, const ProcessorState::CallbackState new_state,
+                     envoy::service::ext_proc::v3::ProcessingRequest& req);
 
   void sendTrailers(ProcessorState& state, const Http::HeaderMap& trailers);
+  bool inHeaderProcessState() {
```

Comment:
Thanks for the comments.

The goal of this check is to ignore mode override when receiving none-header response messages.
send_headers_ is a filter configuration, it will always be true in the filter lifetime if header mode is SEND.  
For example, if ext_proc is configured with header SEND and boy STREAMED,  when receiving the STREAMED body response, send_headers_ will be true. That make it not applicable here.

Reply:
I see. Then, should we move whole mode override logic/code to `handleHeaderResponse`?
That is the natural place representing header response handling and we also don't need this check anymore.

Reply:
There are two issues to move inside handleHeadersResponse:

1) If we want to do it that way, need to pass in the ProcessResponse, instead of request_headers/response_headers, since mode_override is outside of the HeadersResponse.
2)  Need to add in logic in handleHeadersResponse() to retrieve either response->request_headers() or response->response_headers().  I would rather keep that logic outside in onReceiveMessage() which it is now.

Reply:
It might be doable to add a separate mode_override parameter into handleHeadersResponse().  I do not have strong opinion in this way or the current approach.

Reply:
Will below work?
Move `decoding_state_.setProcessingMode(response->mode_override())` before line#639 and move `encode` before line 642.   Or something similar to that like checking `response->response_case()`

Keeping it in this file SGTM, but I feel `inHeaderProcessState` is redundant here because we know what kinds of response we are handling.

Reply:
BTW, the "response" gives us the response type, but it doesn't tell us the ext_proc filter state.  So, even handleHeadersResponse still need to check the state as here:
https://github.com/envoyproxy/envoy/blob/186cabc53948a89be53e653c78b6ed749495204c/source/extensions/filters/http/ext_proc/processor_state.cc#L86

Reply:
I think handleHeadersResponse is for both INBOUND and OUTBOUND direction, rather than particular state. It is same as `onReceiveMessage` here. As long as `handleHeadersResponse` is called, we are fine here.  

From code, check seems for spurious message, though it will be safe to override mode in inside check.

Reply:
Yeah, that's right for the first part. For the spurious message check, it is trying to protect Envoy from misbehaved ext_proc server.